store/src/java/com/zimbra/cs/service/mail/RestoreContacts.java

```diff
@@ -68,85 +59,55 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
-                    HttpServletRequest httpRequest = null;
-                    if (servReq instanceof HttpServletRequest) {
-                        httpRequest = (HttpServletRequest) servReq;
-                        realm = httpRequest.isSecure() ? "https://" : "http://";
-                    }
                     if(resolve == null) {
                         resolve = Resolve.reset;
                     }
-                    String url = realm + getLocalHost() + "/service/home/"
-                        + mbox.getAccount().getName() + "/?" + UserServlet.QP_FMT + "=tgz&"
+                    String CONTACT_RES_URL = "//?" + UserServlet.QP_FMT + "=tgz&"
                         + UserServlet.QP_TYPES + "=contact&" + MimeConstants.P_CHARSET + "=UTF-8";
-                    CookieStore cookieStore = new BasicCookieStore();
                     if (resolve != Resolve.ignore) {
-                        url = url + "&" + MailConstants.A_CONTACTS_RESTORE_RESOLVE + "=" + resolve;
-                    }
-                    AuthToken authToken = octxt.getAuthToken();
-                    BasicClientCookie cookie = null;
-                    try {
-                        cookie = new BasicClientCookie(ZimbraCookie.COOKIE_ZM_AUTH_TOKEN, authToken.getEncoded());
-                        cookie.setPath("/");
-                        cookie.setDomain(mbox.getAccount().getDomainName());
-                        cookieStore.addCookie(cookie);
-                    } catch (AuthTokenException e) {
-                        throw ServiceException.FAILURE("Failed to get authentication token", e);
+                        CONTACT_RES_URL = CONTACT_RES_URL + "&"
+                            + MailConstants.A_CONTACTS_RESTORE_RESOLVE + "=" + resolve;
                     }
                     File file = new File(FileBlobStore.getBlobPath(mbox, doc.getId(),
                         doc.getSavedSequence(), Short.valueOf(doc.getLocator())));
                     if (!file.exists()) {
                         throw ServiceException
                             .INVALID_REQUEST("File does not exist: " + contactBackupFileName, null);
                     }
-                    ZimbraLog.contact.debug("Backup file found. Restoring contacts in %s.",
+                    Account account = mbox.getAccount();
+                    ZimbraAuthToken token = new ZimbraAuthToken(account);
+                    ZMailbox.Options zoptions = new ZMailbox.Options(token.toZAuthToken(),
+                        AccountUtil.getSoapUri(account));
+                    zoptions.setNoSession(true);
+                    zoptions.setTargetAccount(account.getId());
+                    zoptions.setTargetAccountBy(AccountBy.id);
+                    ZMailbox zmbx = ZMailbox.getMailbox(zoptions);
+                    try {
+                        InputStream is = new FileInputStream(file);
+                        zmbx.postRESTResource(CONTACT_RES_URL, is, true, file.length(), null,
```

Comment:
postRESTResource closes the is? if not should close the stream in finally block.

Reply:
Yes, postRESTResource closes input stream. Third parameter is set to true for that.